### ISSUE TITLE
rpi0 bcm2835 need downgrade to 1.9.1 to work on bettercap but system need 1.10.1 to run

### ISSUE THREAD

**Issue by V0r-T3x** [the-tcpdump-group/libpcap#1041]
**libpcap version:**
tcpdump version 4.99.0
libpcap version 1.10.0 (with TPACKET_V3)
OpenSSL 1.1.1k  25 Mar 2021

**Operating system:**
Linux ALOA 4.14.80-Re4son+ # 1 Thu Feb 6 15:03:43 CET 2020 armv6l GNU/Linux

**configure flags:**
no special configuration applied

**Statement of the problem and how this   ##happen:**
I play with the p4wnp1 aloa and I installed bettercap. I could use the native physical Broadcom bcm2835 chipset from the raspberry Pi zero W to scan network with bettercap. After an update of the system, the wlan0 could not use it anymore. in bettercap i had this error message:
`[err] error while setting interface wlan0 in monitor mode: Cannot set rfmon for this handle`

i found a report post on the bettercap caplet issues page about the problem. 
[Bettercap: Cannot set rfmon for this handle AWUS1900 #819](https://github.com/bettercap/bettercap/issues/819)

Its a problem with the new version of the libpcap (1.10.0). I tried to downgrade (1.9.1-4) and the problem was solve. This create an other huge problem in my system. It seem the apt package need the libpcap 1.10.1 to work... then if I keep the 1.9.1-4, my wifi chipset work but i cannot install anything on my system and if i keep, and vice versa. Is anyone can help me to find how I can keep the 1.10.1 and  make the bcm2835 work again?

**Comment by infrastation** [comment]
Is it true in this case that people that can git-bisect the fault don't have the hardware and people that have cannot?

**Comment by infrastation** [comment]
If anybody is willing to work on this issue (reproducing it seems to require Pi Zero W, which isn't expensive), please step in, otherwise it will be closed in 7 days.

**Comment by darkr4y** [comment]
**Same error**
when use rpi OS latest version (debian bullseye)
apt-cache policy libpcap0.8 will show following output:
```
libpcap0.8:
  Installed: 1.10.0-2
  Candidate: 1.10.0-2
  Version table:
 *** 1.10.0-2 500
        500 http://mirrors.xxx.com/raspbian/raspbian bullseye/main armhf Packages
        100 /var/lib/dpkg/status
```
If I want to follow up and fix this error, how do I start?

**Comment by darkr4y** [comment]
Try to make the two version of libpcap's .so file co-exists
* download old version from http://old.kali.org/kali/pool/main/libp/libpcap/
* `dpkg -X` to extract the deb
```
tree libpcap_extracted/
libpcap_extracted/
└── usr
    ├── lib
    │   └── arm-linux-gnueabihf
    │       ├── libpcap.so.0.8 -> libpcap.so.1.9.1
    │       └── libpcap.so.1.9.1
    └── share
        ├── doc
        │   └── libpcap0.8
        │       ├── changelog.Debian.gz
        │       ├── changelog.gz
        │       ├── copyright
        │       ├── CREDITS.gz
        │       ├── README.Debian
        │       └── README.md
        └── man
            └── man7
                └── pcap-filter.7.gz

8 directories, 9 files
```
the newest version 
```
ls -al /usr/lib/arm-linux-gnueabihf/libpcap*
-rw-r--r-- 1 root root 346020 Jan  5  2021 /usr/lib/arm-linux-gnueabihf/libpcap.a
lrwxrwxrwx 1 root root     14 Jan  5  2021 /usr/lib/arm-linux-gnueabihf/libpcap.so -> libpcap.so.0.8
lrwxrwxrwx 1 root root     17 Jan  5  2021 /usr/lib/arm-linux-gnueabihf/libpcap.so.0.8 -> libpcap.so.1.10.0
-rw-r--r-- 1 root root 268436 Jan  5  2021 /usr/lib/arm-linux-gnueabihf/libpcap.so.1.10.0
```

**Comment by darkr4y** [comment]
I also tried the aircrack-ng
```
sudo -i
ip link set wlan0 down
iw dev wlan0 set type monitor
ip link set wlan0 up
airmon-ng start wlan0
airodump-ng wlan0mon
```
it runs very well as always.
Still not sure where the key problem :(

**Comment by darkr4y** [comment]
I think I may have found the problem.
There may be bugs in higher versions of libpcap,but during my testing, I confused the rpiOS and kali deb packages, resulting in the `Illegal instruction` error above.
replace the download url to https://mirrors.xxxx.com/raspbian/raspbian/pool/main/libp/libpcap/libpcap0.8_1.8.1-3%2Bdeb9u1_armhf.deb 
Then repeat the steps for setting `LD_LIBRARY_PATH`, now two version of libpcap can co-exists.

**Comment by infrastation** [comment]
Do you have the hardware and the OS where this problem reproduces?

**Comment by darkr4y** [comment]
> Do you have the hardware and the OS where this problem reproduces?

yep

I have Raspi Zero W v1.1
and the os is RaspiOS which used to be called Raspbian(the lastest version based on Debian Bullseye)

**Comment by infrastation** [comment]
As far as the problem description goes, the bug is not present in libpcap 1.9.1, but is present in 1.10.0. If you know how to do git bisection, it would help if you could identify the commit that introduced the problem.

**Comment by darkr4y** [comment]
OK got
I will invest the commit history
and git diff the code.
Reply soon 👌🏻

**Comment by darkr4y** [comment]
Got
I'll use quick dirty test code and `git bitsect` to find out "bad revision"

**Comment by infrastation** [comment]
Thank you trying to bisect the problem. From your comments I understand the problem does not reproduce as originally described, is this correct?

**Comment by darkr4y** [comment]
yep

**Comment by infrastation** [comment]
`libpcap-1.10-bp` is commit 8c6c412 and `libpcap-1.10.0` is commit ef0762f, which is the next commit and which does not change any C code. So these results don't make sense. Let me suggest to put dynamic linking complexity out of the problem space, in that you can compile tcpdump to use static libpcap, and then run tcpdump with `--monitor-mode` to see if the function call fails. Alternatively, you could provide remote SSH access to the board.

**Comment by infrastation** [comment]
For reference, on RPI3B rev. 1.2 it is impossible to reproduce this issue because the interface does not support monitor mode (tested with `2022-01-28-raspios-bullseye-armhf-lite.img` and the latest master branches):
```
tcpdump -i wlan0 --monitor-mode
tcpdump: wlan0: That device doesn't support monitor mode
```
I wonder if it looks the same on RPI0W.

**Comment by infrastation** [comment]
Thank you for trying. It is important to know the steps to reproduce the issue. Maybe developers of the Linux distribution where this happens would be in a better position to state the problem.

**Comment by infrastation** [comment]
If this problem report does not have exact steps to reproduce in 14 days, it will be closed.

**Comment by infrastation** [comment]
It is time.

**Comment by evilsocket** [comment]
for the record i'm still looking into this, airodump-ng works because it's not using the system libpcap:

```
ldd $(which airodump-ng)                                                                                                                   

        linux-vdso.so.1 (0x00007fff5ffd7000)
        libaircrack-osdep-1.6.0.so => /lib/x86_64-linux-gnu/libaircrack-osdep-1.6.0.so (0x00007f1180414000)
        libpcre.so.3 => /lib/x86_64-linux-gnu/libpcre.so.3 (0x00007f118039e000)
        libgcrypt.so.20 => /lib/x86_64-linux-gnu/libgcrypt.so.20 (0x00007f1180260000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f1180037000)
        libnl-3.so.200 => /lib/x86_64-linux-gnu/libnl-3.so.200 (0x00007f1180014000)
        libnl-genl-3.so.200 => /lib/x86_64-linux-gnu/libnl-genl-3.so.200 (0x00007f1180009000)
        libgpg-error.so.0 => /lib/x86_64-linux-gnu/libgpg-error.so.0 (0x00007f117ffe3000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f118045c000)
```

**Comment by infrastation** [comment]
Reopening, I now have the hardware where this problem supposedly reproduces (RPI0W, BCM2835). Pending some tests.

**Comment by infrastation** [comment]
A Raspberry Pi Zero W rev. 1.1 running `2025-05-13-raspios-bookworm-armhf.img.xz` with all current updates does not seem to support 802.11 monitor mode in the first place, which is the same behaviour I have observed with models 4B and 5. At least in this particular setup this does not look like a regression in libpcap.
```
# uname -a
Linux rpi0w 6.12.25+rpt-rpi-v6 #1 Raspbian 1:6.12.25-1+rpt1 (2025-04-30) armv6l GNU/Linux
# ip link set wlan0 down
# iw dev wlan0 set type monitor
command failed: Operation not supported (-95)
# iw phy phy0 interface add mon0 type monitor
command failed: Operation not supported (-95)
```
Attempts to enable monitor mode via libpcap consistently fail too:
* libpcap from Debian packages (without libnl):
  ```
  # tcpdump --version
  tcpdump version 4.99.3
  libpcap version 1.10.3 (with TPACKET_V3)
  OpenSSL 3.0.16 11 Feb 2025
  # tcpdump -ni wlan0 -I
  tcpdump: wlan0: That device doesn't support monitor mode
  ```
* locally compiled libpcap 1.9.1 with libnl:
  ```
  # tcpdump --version
  tcpdump version 5.0.0-PRE-GIT
  libpcap version 1.9.1 (with TPACKET_V3)
  32-bit build, 32-bit time_t
  # tcpdump -ni wlan0 -I
  tcpdump: wlan0: nl_wait_for_ack failed adding mon0 interface: Operation not supported
  # opentest -i wlan0 -I
  opentest: wlan0: Generic error
  (wlan0: nl_wait_for_ack failed adding mon0 interface: Operation not supported)

  ```
* libpcap 1.9.1 without libnl:
  ```
  # tcpdump --version
  tcpdump version 5.0.0-PRE-GIT
  libpcap version 1.9.1 (with TPACKET_V3)
  32-bit build, 32-bit time_t
  # tcpdump -ni wlan0 -I
  tcpdump: wlan0: That device doesn't support monitor mode
  # opentest -i wlan0 -I
  opentest: wlan0: That device doesn't support monitor mode
  (That device doesn't support monitor mode)
  ```

So either this problem does not exist, or it needs a different setup to reproduce.

**Comment by infrastation** [comment]
After some more research I managed to establish the following.

The original problem statement refers to P4wnP1 A.L.O.A., which is a modification of Kali Linux. P4wnP1 among many other things includes Re4son-Kernel for Raspberry Pi (4.14.80), which in turn includes nexmon firmware for Broadcom 802.11 chips, including the BCM43430A1 on RPI0W. Using this image from 2020, it is indeed possible to change `wlan0` mode to monitor (in which case it does not deliver any packets) or to add a monitor interface to the onboard 802.11 PHY (in which case the `mon0` interface presents `DLT_IEEE802_11_RADIO` and correctly delivers packets with 802.11 headers). So supposedly libpcap cannot drive this case properly. The workaround done in Bettercap, as far as I understand, boils to keep the interface down whilst monitor mode is being activated.

That said, this image does not include all the packages required to build libpcap for debugging, and no newer builds exist, supposedly because the Re4son-Kernel has not made any releases after 2019. Kali Linux [mentions](https://www.kali.org/docs/arm/raspberry-pi-zero-w-p4wnp1-aloa/) building P4wnP1 images, but they are absent from the available downloads, they also mention nexmon integration as a work in progress. A possible reason why it has not been done is that nexmon README file lists BCM43430A1 as supported in Debian 8 and 9 only. For what it's worth, out of the box the most recent Kali release (2025.1c) allows both to change `wlan0` mode and to create a monitor device same as the above, but neither of that venues delivers any packets in the packet capture.

One potential reason to investigate this further is that nexmon also supports RPI3B, RPI3B+, RPI4B and RPI5 for kernels that are a bit newer. Another potential reason is that some other, unmodified 802.11 devices could have the same quirk and require the same workaround.

The main reason I started to look into this problem was to verify the problem space of #1508, and so far I have not seen an obvious connection.
